### PR TITLE
feat: add token exchange rate for udt (Issue-1130)

### PR DESCRIPTION
**Pull request by Naupio**
## What problem does this PR solve?
link issue #1130 
add token exchange rate for udt
## Check List
check udt object(must with valid symbol like: "CKB", "BTC", "ETH", etc...) with `token exchange rate` field
#### Test
- [x] token exchange rate subfield testing code
#### Task
 - none

## addition
- [x] compatible with uan format symbol for fetching price
- [x] price show condition: only show type with `native` record for `udt` object and show `eth-bridge-celer-mapping-godwoken-native-token` and `ckb-bridge-mapping-godoweken-native-token`(if udt record has uan fields( bridge token), show the price)

### CODE REVIEW

File: lib/godwoken_explorer/graphql/resolovers/udt.ex [hunk 1]
```diff
@@ -53,6 +54,25 @@ defmodule GodwokenExplorer.Graphql.Resolvers.UDT do
     {:ok, return}
   end
 
+  def token_exchange_rate(%UDT{} = udt, _args, _resolution) do
+    symbol = udt.symbol
+
+    if symbol do
```
Comment (Keith-CY) [inline lib/godwoken_explorer/graphql/resolovers/udt.ex:60]: Is `symbol` enough to locate the official token? Say I have minted a token with the symbol `USDT`, will the price be returned along?

Comment (Naupio) [reply lib/godwoken_explorer/graphql/resolovers/udt.ex:60]: yes, if you minted a token witch symbol also name 'USDT', it will return the price too.

**Comment by zmcNotafraid** [comment]
pCKB.gw|gb.ckb

**Comment by Naupio** [comment]
- [x] compatible with uan format symbol for fetching price
- [x] price show condition: only show type with `native` record for `udt` object and show `eth-bridge-celer-mapping-godwoken-native-token` and `ckb-bridge-mapping-godoweken-native-token`(if udt record has uan fields( bridge token), show the price)

File: lib/godwoken_explorer/chain/cache/token_exchange_rate.ex [hunk 2]
```diff
@@ -91,15 +105,19 @@ defmodule GodwokenExplorer.Chain.Cache.TokenExchangeRate do
...
-    case Source.fetch_exchange_rates_for_token(symbol) do
-      {:ok, [rates]} ->
-        rates.usd_value
-
-      {:error, "Could not find coin with the given id"} ->
-        if internal_call?, do: :not_found_coingecko, else: nil
-
-      _ ->
-        nil
+    try do
```
Comment (zmcNotafraid) [inline lib/godwoken_explorer/chain/cache/token_exchange_rate.ex:108]: Why use try catch？

**Review by FrederLu — APPROVED** [review]

File: lib/godwoken_explorer/chain/cache/token_exchange_rate.ex [hunk 3]
```diff
@@ -43,6 +43,20 @@ defmodule GodwokenExplorer.Chain.Cache.TokenExchangeRate do
...
+    if cache_expired?(symbol) do
+      update_cache_by_symbol(symbol)
+    end
+
+    cached_value =
+      symbol
+      |> cache_key()
+      |> fetch_from_cache()
+
+    updated_at = fetch_from_cache("#{cache_key(symbol)}_#{@last_update_key}")
```
[the remark this answers is not in the document]
Comment (Naupio) [reply lib/godwoken_explorer/chain/cache/token_exchange_rate.ex:56]: 一个是 value 一个是 update_at, 两条记录在 ets 里面

[the remark this answers is not in the document]
Comment (Naupio) [reply lib/godwoken_explorer/chain/cache/token_exchange_rate.ex:56]: > 上边不是已经读取了cache?

一个是 value 一个是 update_at, 两条记录在 ets 里面

File: lib/godwoken_explorer/chain/cache/token_exchange_rate.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (Naupio) [reply lib/godwoken_explorer/chain/cache/token_exchange_rate.ex:108]: GodwokenExplorer.ExchangeRates.Source.CoinMarketCap.format_data callback function will sometime raise error with `symbol: token_properties && String.upcase(token_properties["symbol"])` when  token_properties not nil and token_properties["symbol"] not exist

File: lib/godwoken_explorer/graphql/resolovers/udt.ex [hunk 4]
```diff
@@ -53,6 +54,27 @@ defmodule GodwokenExplorer.Graphql.Resolvers.UDT do
     {:ok, return}
   end
 
+  def token_exchange_rate(%UDT{} = udt, _args, _resolution) do
+    symbol = udt.symbol
+    uan = udt.uan
+
+    if not is_nil(symbol) and not is_nil(uan) do
```
[the remark this answers is not in the document]
Comment (Naupio) [reply lib/godwoken_explorer/graphql/resolovers/udt.ex:61]: fixed with https://github.com/Magickbase/godwoken_explorer/pull/1158/commits/44cad500f53a9b49ca49611e6d6614284ca175f2

File: lib/godwoken_explorer/chain/cache/token_exchange_rate.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (zmcNotafraid) [reply lib/godwoken_explorer/chain/cache/token_exchange_rate.ex:108]: 你这里肯定要保证传进去的不是nil啊

**Review by zmcNotafraid — APPROVED** [review]

File: lib/godwoken_explorer/chain/cache/token_exchange_rate.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (Naupio) [reply lib/godwoken_explorer/chain/cache/token_exchange_rate.ex:108]: 这里触发是类似这种 `GodwokenExplorer.ExchangeRates.Source.fetch_exchange_rates_for_token "pCKB"`
会产生这种错误
```
** (FunctionClauseError) no function clause matching in String.upcase/2    
    
    The following arguments were given to String.upcase/2:
    
        # 1
        nil
    
        # 2
        :default
    
    Attempted function clauses (showing 4 out of 4):
    
        def upcase("", _mode)
        def upcase(string, :default) when is_binary(string)
        def upcase(string, :ascii) when is_binary(string)
        def upcase(string, mode) when is_binary(string) and mode === :greek or mode === :turkic
    
    (elixir 1.14.2) lib/string.ex:831: String.upcase/2
    (godwoken_explorer 0.1.0) lib/godwoken_explorer/exchange_rates/source/coin_market_cap.ex:46: GodwokenExplorer.ExchangeRates.Source.CoinMarketCap.format_data/1
    (godwoken_explorer 0.1.0) lib/godwoken_explorer/exchange_rates/source.ex:34: GodwokenExplorer.ExchangeRates.Source.fetch_exchange_rates_request/3
```

**Pull request merged**